<header>
   <p  style='font-size:36px;font-family:Arial; color:#F0F0F0; background-color: #00233c; padding-left: 20pt; padding-top: 20pt;padding-bottom: 10pt; padding-right: 20pt;'>
       HCA Healthcare Peer Facilities Clustering
  <br>
       <img id="teradata-logo" src="https://storage.googleapis.com/clearscape_analytics_demo_data/DEMO_Logo/teradata.svg" alt="Teradata" style="width: 125px; height: auto; margin-top: 20pt;">
    </p>
</header>


<p style = 'font-size:16px;font-family:Arial;color:#00233C'>A peer facility would be a facility of similar size, volume, and case mix. Peer Group Clusters allows the business to compare like-facilities together across several metrics, including revenue, expenses, outcomes, etc. In this demo, you will learn an efficient way to group acute care facilities into peer clusters with ClearScape Analytics. A peer facility would be a facility of similar size, volume, and case mix.</p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>Grouping Columns:

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>
    <b>count_fte</b>: Total number of full-time employees as of the period end date. <br>
 <b>count_bed</b>: Total number of beds in service as of the period end date.<br>
 <b>count_admissions</b>: Total adjusted admissions as of the period end date. Adjusted Admissions is all inpatient admissions adjusted by an outpatient factor. Representative of total inpatient and outpatient volume.<br>
 <b>score_case_mix</b>: An indicator of inpatient acuity. Total DRG Weight / Total inpatient admissions as of the period end date.<br>
 <b>sum_net_revenue</b>: Total net revenue as of the period end date.<br>
 <b>sum_opex</b>: Total operating expenses as of the period end date.</p>
</p>


<p style = 'font-size:18px;font-family:Arial;color:#00233C'><b>1. Connect to Vantage</b></p>
<p style = 'font-size:16px;font-family:Arial;color:#00233C'>You will be prompted to provide the password. Enter your password, press Enter, then use down arrow to go to next cell.</p>

In [1]:
%connect local

Password: ········


Success: 'AzureCloud' connection established and activated for user 'db255096', with default database 'db255096'



<p style = 'font-size:18px;font-family:Arial;color:#00233C'><b>2. Data Preparation</b></p>
<p style = 'font-size:16px;font-family:Arial;color:#00233C'>

In [2]:
select top 5* from features

entity_id entity_name special_facility Exclude_Facility measure_dt count_fte count_beds count_admissions score_case_mix sum_net_revenue sum_opex 1 26109 METHODIST TEXSAN HOSPITAL 0 0 2023-12-31 518 120 10819 1.88 165664371 123692956 2 27400 THE MEDICAL CENTER AURORA 0 0 2023-12-31 1465 389 30626 1.78 420474649 343874458 3 8298 CORPUS CHRISTI MEDICAL CENTER 0 0 2023-12-31 1656 421 39350 1.51 490845548 376009568 4 25325 SPOTSYLVANIA REGIONAL MEDICAL CENTER 0 0 2023-12-31 522 127 19552 1.32 182678920 124783461 5 31767 GREENVIEW REGIONAL HOSPITAL 0 0 2023-12-31 438 118 14437 1.54 143532991 109108076

<p style = 'font-size:16px;font-family:Arial;color:#00233C'><b><a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Feature-Engineering-Transform-Functions/Antiselect'>Antiselect</a> </b> returns all columns except those specified in the Exclude syntax element.</p>

In [6]:
drop view v_features;

Success: 15 rows affected

In [7]:
--Create View to that can be used to filter out special or excluded facilities with missing
create volatile table v_features AS (

select * from Antiselect (
        on (SELECT hf.*
        FROM features hf
        JOIN original_cluster hu ON hf.entity_id = hu.entity_id
        WHERE hu.cluster_id IS NOT NULL
        AND hf.Exclude_Facility < 1
        AND hf.special_facility < 1)  
        USING
  Exclude ('special_facility', 'Exclude_Facility')
) as dt)with data ON COMMIT PRESERVE ROWS;

Success: 0 rows affected

In [8]:
select top 5* from v_features;

entity_id entity_name measure_dt count_fte count_beds count_admissions score_case_mix sum_net_revenue sum_opex 1 32905 PARKLAND MEDICAL CENTER 2023-12-31 430 85 12513 1.44 159773914 113260281 2 30997 HCA FLORIDA OAK HILL HOSPITAL 2023-12-31 1275 350 32348 1.52 370906368 246575422 3 37938 HCA FLORIDA LAKE CITY HOSPITAL 2023-12-31 565 113 24559 1.38 168408021 103403864 4 30932 HCA FLORIDA LAWNWOOD HOSPITAL 2023-12-31 1637 432 34466 1.72 531541178 357143367 5 31003 DOCTORS HOSPITAL OF AUGUSTA 2023-12-31 1501 331 25122 1.88 493507285 365612369

<p style = 'font-size:16px;font-family:Arial;color:#00233C'> <b> 2.1 Checking data demographics  </b> </p>
<p style = 'font-size:16px;font-family:Arial;color:#00233C'><b><a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Data-Exploration-Functions/TD_ColumnSummary'>TD_ColumnSummary</a> </b>function  displays Column name, datatype and other demographics like count of NULLs etc for each specified input table column</p>

In [9]:
SELECT * FROM TD_ColumnSummary (
 ON v_features AS InputTable
 USING
 TargetColumns ('[:]')
) AS dt;

ColumnName Datatype NonNullCount NullCount BlankCount ZeroCount PositiveCount NegativeCount NullPercentage NonNullPercentage 1 count_fte INTEGER 165 2 1 164 0 1.1976047904191616 98.80239520958084 2 sum_net_revenue BIGINT 165 2 0 165 0 1.1976047904191616 98.80239520958084 3 count_admissions INTEGER 165 2 2 163 0 1.1976047904191616 98.80239520958084 4 sum_opex BIGINT 167 0 2 165 0 0 100 5 measure_dt DATE 167 0 0 100 6 score_case_mix DECIMAL(3,2) 165 2 2 163 0 1.1976047904191616 98.80239520958084 7 count_beds INTEGER 165 2 2 163 0 1.1976047904191616 98.80239520958084 8 entity_name VARCHAR(255) CHARACTER SET LATIN 167 0 0 0 100 9 entity_id INTEGER 167 0 0 167 0 0 100

<hr style="height:1px;border:none;background-color:#00233C;">
<p style = 'font-size:16px;font-family:Arial;color:#00233C'><b>2.2 Gather statistics using TD_UnivariateStatistics</b></p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>Univariate analysis is the simplest form of analyzing data. Its significant purpose is to describe; It takes data, summarizes it, and finds patterns in the data.
    <br>
    <a href = 'https://docs.teradata.com/search/all?query=TD_UnivariateStatistics&content-lang=en-US'>TD_UnivariateStatistics</a> displays descriptive statistics for each specified numeric input table column.
</p>

In [10]:
SELECT *
FROM TD_UnivariateStatistics (
ON v_features  AS InputTable
USING
TargetColumns('count_fte' , 'count_beds' ,'count_admissions' ,'score_case_mix','sum_net_revenue' ,'sum_opex')
STATS( 
      'MEAN',
      'MEDIAN',
      'MODE',           
      'ZERO VALUES COUNT',
      'PERCENTILES',
      'MINIMUM',
      'MAXIMUM'
        )
        ) as dt order by 1,2;

ATTRIBUTE StatName StatValue 1 count_admissions MAXIMUM 78804 2 count_admissions MEAN 21993.951515151515 3 count_admissions MEDIAN 19048 4 count_admissions MINIMUM 0 5 count_admissions MODE 0 6 count_admissions PERCENTILES(1) 821.1200000000001 7 count_admissions PERCENTILES(10) 8235.6 8 count_admissions PERCENTILES(25) 11554 9 count_admissions PERCENTILES(5) 5561.8 10 count_admissions PERCENTILES(50) 19048 11 count_admissions PERCENTILES(75) 29288 12 count_admissions PERCENTILES(90) 39274.4 13 count_admissions PERCENTILES(95) 46549.39999999999 14 count_admissions PERCENTILES(99) 66328.71999999997 15 count_admissions ZERO VALUES COUNT 2 16 count_beds MAXIMUM 827 17 count_beds MEAN 239.73333333333332 18 count_beds MEDIAN 206 19 count_beds MINIMUM 0 20 count_beds MODE 97 21 count_beds PERCENTILES(1) 6.400000000000001 22 count_beds PERCENTILES(10) 57.400000000000006 23 count_beds PERCENTILES(25) 110 24 count_beds PERCENTILES(5) 31.400000000000006 25 count_beds PERCENTILES(50) 206 26 count_beds PERCENTILES(75) 331 27 count_beds PERCENTILES(90) 444.4 28 count_beds PERCENTILES(95) 511.5999999999999 29 count_beds PERCENTILES(99) 809.4799999999998 30 count_beds ZERO VALUES COUNT 2 31 count_fte MAXIMUM 4051 32 count_fte MEAN 980.2242424242424 33 count_fte MEDIAN 861 34 count_fte MINIMUM 0 35 count_fte MODE 462 36 count_fte PERCENTILES(1) 44.84 37 count_fte PERCENTILES(10) 270.6 38 count_fte PERCENTILES(25) 426 39 count_fte PERCENTILES(5) 173.4 40 count_fte PERCENTILES(50) 861 41 count_fte PERCENTILES(75) 1324 42 count_fte PERCENTILES(90) 1781.8 43 count_fte PERCENTILES(95) 2197.599999999998 44 count_fte PERCENTILES(99) 3578.4799999999896 45 count_fte ZERO VALUES COUNT 1 46 score_case_mix MAXIMUM 2.75 47 score_case_mix MEAN 1.5578181818181815 48 score_case_mix MEDIAN 1.54 49 score_case_mix MINIMUM 0 50 score_case_mix MODE 1.33 51 score_case_mix PERCENTILES(1) 0.6656000000000002 52 score_case_mix PERCENTILES(10) 1.31 53 score_case_mix PERCENTILES(25) 1.4 54 score_case_mix PERCENTILES(5) 1.192 55 score_case_mix PERCENTILES(50) 1.54 56 score_case_mix PERCENTILES(75) 1.73 57 score_case_mix PERCENTILES(90) 1.88 58 score_case_mix PERCENTILES(95) 1.9779999999999998 59 score_case_mix PERCENTILES(99) 2.149599999999998 60 score_case_mix ZERO VALUES COUNT 2 61 sum_net_revenue MAXIMUM 1.371232053e+09 62 sum_net_revenue MEAN 3.202118821757576e+08 63 sum_net_revenue MEDIAN 2.74842739e+08 64 sum_net_revenue MINIMUM 2.77553e+06 65 sum_net_revenue MODE 2.77553e+06 66 sum_net_revenue PERCENTILES(1) 1.065854096e+07 67 sum_net_revenue PERCENTILES(10) 7.363873280000001e+07 68 sum_net_revenue PERCENTILES(25) 1.21607346e+08 69 sum_net_revenue PERCENTILES(5) 4.41636144e+07 70 sum_net_revenue PERCENTILES(50) 2.74842739e+08 71 sum_net_revenue PERCENTILES(75) 4.28004422e+08 72 sum_net_revenue PERCENTILES(90) 6.297824834e+08 73 sum_net_revenue PERCENTILES(95) 7.603527953999996e+08 74 sum_net_revenue PERCENTILES(99) 1.1883711930799959e+09 75 sum_net_revenue ZERO VALUES COUNT 0 76 sum_opex MAXIMUM 9.87982056e+08 77 sum_opex MEAN 2.2574740194610777e+08 78 sum_opex MEDIAN 1.92259193e+08 79 sum_opex MINIMUM 0 80 sum_opex MODE 0 81 sum_opex PERCENTILES(1) 1.7224944000000004e+06 82 sum_opex PERCENTILES(10) 5.3591795400000006e+07 83 sum_opex PERCENTILES(25) 8.9324818e+07 84 sum_opex PERCENTILES(5) 3.20926855e+07 85 sum_opex PERCENTILES(50) 1.92259193e+08 86 sum_opex PERCENTILES(75) 3.040124e+08 87 sum_opex PERCENTILES(90) 4.394243896000001e+08 88 sum_opex PERCENTILES(95) 5.1526620179999995e+08 89 sum_opex PERCENTILES(99) 8.336489842000006e+08 90 sum_opex ZERO VALUES COUNT 2

<hr style="height:1px;border:none;background-color:#00233C;">
<p style = 'font-size:18px;font-family:Arial;color:#00233C'><b>2.3 Scale Transform </b></p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>Many machine learning models, especially those that rely on distance calculations (e.g., K-Means, K-Nearest Neighbors, Support Vector Machines, and Principal Component Analysis), assume that all features contribute equally. If one feature has a much larger scale than others, it can dominate the model’s behavior.

<p style = 'font-size:16px;font-family:Arial;color:#00233C'><a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Feature-Engineering-Transform-Functions/TD_ScaleFit'>TD_ScaleFit</a> outputs a table of statistics to input to TD_ScaleTransform, which scales specified input table columns. TD_ScaleFit accepts input data in dense and sparse format.
    <br>
    <a href = 'https://docs.teradata.com/search/all?query=TD_UnivariateStatistics&content-lang=en-US'>TD_ScaleTransform</a> scales specified input table columns, using TD_ScaleFit output. TD_ScaleTransform accepts input data in dense and sparse format.
</p>


In [11]:
drop table features_scaled;

Success: 20 rows affected

In [13]:
--Calculate Scaling parameters
create volatile table features_scaled as (
SELECT * FROM TD_ScaleFit(
    ON v_features AS InputTable
    --OUT TABLE OutputTable(features_scaled)
    USING
        TargetColumns('count_fte' , 'count_beds' ,'count_admissions' ,'score_case_mix','sum_net_revenue' ,'sum_opex')
        ScaleMethod('range')
        MissValue('zero')
) as dt)with data ON COMMIT PRESERVE ROWS;


Success: 0 rows affected

In [15]:
drop table v_features_scaled

Success: 21 rows affected

In [16]:
create volatile table v_features_scaled as (

SELECT hf.entity_id, hs.count_fte, hs.count_beds, hs.count_admissions, hs.score_case_mix, hs.sum_net_revenue, hs.sum_opex 
FROM TD_scaleTransform (
  ON v_features AS InputTable
  ON features_scaled AS FitTable DIMENSION
  USING
  Accumulate ('entity_id')
) AS hs
	JOIN v_features hf on (hf.entity_id = hs.entity_id)
) with data ON COMMIT PRESERVE ROWS;

Success: 0 rows affected

In [17]:
select * from v_features_scaled

entity_id count_fte count_beds count_admissions score_case_mix sum_net_revenue sum_opex 1 39233 0.4401382374722291 0.4848851269649335 0.4606999644688087 0.5818181818181819 0.41647986137737164 0.40546270812027785 2 26890 0.04196494692668477 0.06408706166868199 0.07939952286685956 0.49090909090909096 0.029989674724945573 0.03676272841133463 3 9727 0.42754875339422366 0.5344619105199516 0.35595908837114865 0.6872727272727273 0.43068909687238927 0.4853200339905768 4 34293 0.31695877561096025 0.34582829504232165 0.2692502918633572 0.7309090909090908 0.2866084938819792 0.28227387360565587 5 30941 0.27993088126388543 0.3857315598548972 0.3912745545911375 0.5236363636363636 0.23687538372748143 0.23057050441005175 6 26920 0.3376943964453221 0.37122128174123337 0.3171285721536978 0.5672727272727273 0.30708693037520784 0.33625807066277325 7 9722 0.2979511231794619 0.32769044740024184 0.3025100248718339 0.5927272727272727 0.3578935748183795 0.3089273232711425 8 31054 0.15971365095038262 0.16807738814993953 0.23383330795391097 0.48363636363636364 0.12179145643197026 0.13036132611704032 9 27620 0.05726980992347568 0.03506650544135429 0.12800111669458403 0.45454545454545453 0.042954711393487215 0.05234215711302352 10 26110 0.3268328807701802 0.4026602176541717 0.3725952997309781 0.52 0.275964822157525 0.2909644373136267 11 34415 0.20093803999012588 0.2430471584038694 0.2547586416933151 0.5709090909090909 0.22758665238208667 0.1854124848599477 12 35940 0.182670945445569 0.2889963724304716 0.2277295568752855 0.5272727272727272 0.12310130805668278 0.1404775270533861 13 26516 0.09553196741545297 0.05804111245465538 0.09782498350337546 0.6181818181818182 0.07530833626637622 0.08563890152272158 14 30908 0.24389039743273266 0.2974607013301088 0.2712171970965941 0.6690909090909091 0.236057079323082 0.22574568297625033 15 30916 0.5033325104912367 0.585247883917775 0.6718440688289935 0.5781818181818182 0.4782426573299326 0.4380661221239832 16 37333 0.1083683041224389 0.16686819830713423 0.10632708999543171 0.5454545454545454 0.05919464859754262 0.07817605950527506 17 31202 0.27474697605529497 0.3591293833131802 0.23456931120247704 0.6545454545454545 0.29245895377269504 0.24699895055583884 18 31201 0.10120957788200444 0.1354292623941959 0.13836861073041978 0.5563636363636364 0.09069958008450371 0.09045372581139267 19 31767 0.1081214514934584 0.14268440145102781 0.18320136033703874 0.56 0.10285855533899194 0.11043528102295817 20 27635 0.04616144161935325 0.045949214026602174 0.14028475711892796 0.4509090909090909 0.029402127377633903 0.038387737681745915 21 37939 0 0 0 0 0 0.0026415864378816207 22 34318 0.2695630708467045 0.27448609431680776 0.2457870158875184 0.7672727272727272 0.29645601828155416 0.29280309418899003 23 34372 0.16514440878795358 0.1777509068923821 0.17678036647885895 0.509090909090909 0.14056955319142425 0.13223788246615686 24 34634 0.4895087632683288 0.6082224909310762 0.6272143546012893 0.5563636363636364 0.5563962210000003 0.5181610069646851 25 39385 0.9967909158232535 1 0.8344880970509111 0.6690909090909091 1 0.9762754031233134 26 38318 0.2540113552209331 0.3542926239419589 0.28855134257144305 0.5527272727272727 0.2192843966589065 0.2252676722703555 27 34336 0.49666748950876327 0.48972188633615477 0.47090249225927616 0.6036363636363636 0.45383701020949424 0.45482483135300994 28 38340 0.3216489755615897 0.32889963724304716 0.36136490533475457 0.5381818181818182 0.26550393300291936 0.27255731049431126 29 39309 0.31523080720809676 0.3675937122128174 0.34754580985736766 0.5418181818181819 0.26165881705545424 0.286313367011192 30 27200 0.33349790175265365 0.4655380894800484 0.20125881934927162 0.8072727272727274 0.45455219697907784 0.4086012752442135 31 26960 0.6553937299432239 0.7243047158403869 0.6742678036647886 0.6836363636363636 0.5603193715786029 0.6145947826809518 32 26670 0.19353246112071093 0.24667472793228537 0.2698720877112837 0.6327272727272727 0.17938655622163308 0.17844794743923972 33 30920 0.30165391261416935 0.5683192261

<hr style="height:1px;border:none;background-color:#00233C;">
<p style = 'font-size:18px;font-family:Arial;color:#00233C'><b> 3. Kmeans Clustering</b></p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>Kmeans is an unsupervised machine learning algorithm used for clustering. It partitions a dataset into K clusters, where each data point belongs to the cluster with the nearest mean (centroid), in this case, the HCA healthcare facilities will be grouped into K cluters using Kmeans clustering method.</p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'><a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Model-Training-Functions/TD_KMeans'>TD_KMeans</a> The k-means algorithm groups a set of observations into k clusters in which each observation belongs to the cluster with the nearest mean (cluster centers or cluster centroid).
    <br>
    <a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Model-Scoring-Functions/TD_KMeansPredict'>TD_KMeansPredict</a>  uses the k-means algorithm to predict the target class of unseen or new data.
</p>

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>This algorithm minimizes the objective function, that is, the total Euclidean distance of all data points from the center of the cluster as follows:</p>
<p style = 'font-size:16px;font-family:Arial;color:#00233C'>
    1.Specify or randomly select k initial cluster centroids. <br>
    2.Assign each data point to the cluster that has the closest centroid.<br>
    3.Recalculate the positions of the k centroids.<br>
    4.Repeat steps 2 and 3 until the centroids no longer move.
</p>



In [18]:
drop table KMeans_Model;

Success: 24 rows affected

In [19]:
--perform KMeans Clustering
Select * from TD_KMeans (
    ON v_features_scaled as InputTable
    OUT TABLE ModelTable(KMeans_Model) --KMeans_Model table is used to save the weights of the model
    USING
        IdColumn('entity_id')
        TargetColumns( 'count_fte' , 'count_beds' ,'count_admissions' ,'score_case_mix','sum_net_revenue' ,'sum_opex')
        InitialCentroidsMethod('kmeans++')
        StopThreshold(0.001)
        NumClusters(5)
        --Seed(0)
        MaxIterNum(100)
)as dt;


Success: 0 rows affected

td_clusterid_kmeans count_fte count_beds count_admissions score_case_mix sum_net_revenue sum_opex td_size_kmeans td_withinss_kmeans entity_id td_modelinfo_kmeans 1 0 0.38242409281658846 0.46571083088616333 0.41366500612731766 0.6214025974025974 0.3856629835259401 0.3757373682792589 35 0.9395282497611674 2 1 0.08832739711535065 0.10559682155812745 0.13466448403634337 0.4827012987012987 0.07499661856003924 0.07848927590817936 70 2.046354238326643 3 2 0.9983954579116268 0.996372430471584 0.9172440485254556 0.6763636363636363 0.9990012828489401 0.9881377015616567 2 0.014117770885253123 4 3 0.24302641323130098 0.2929141475211609 0.2779838586873763 0.6045090909090909 0.22680209513678498 0.2288815720353529 50 1.2514009656441658 5 4 0.6209331029375462 0.7500604594921403 0.6412517131110096 0.6352727272727272 0.6184375034032411 0.6092990451033049 10 0.7359083241183697 6 Converged : True 7 Number of Iterations : 13 8 Number of Clusters : 5 9 Total_WithinSS : 4.98730954873560E+00 10 Between_SS : 2.66069968833987E+01 11 Method for InitialCentroids : KMeans++

In [21]:
drop table clusters

Success: 17 rows affected

In [22]:
create volatile table clusters as (

SELECT entity_id, td_clusterid_kmeans as cluster_id, CURRENT_DATE() as cluter_date 
FROM TD_KMeansPredict (
    ON v_features_scaled AS InputTable
    ON KMeans_Model as ModelTable DIMENSION
    USING
        OutputDistance('false')
) as dt) with data ON COMMIT PRESERVE ROWS;

Success: 0 rows affected

In [23]:
select * from clusters

entity_id cluster_id cluter_date 1 39233 0 2025-03-21 2 26890 1 2025-03-21 3 9727 0 2025-03-21 4 34293 3 2025-03-21 5 30941 3 2025-03-21 6 26920 0 2025-03-21 7 9722 3 2025-03-21 8 31054 1 2025-03-21 9 27620 1 2025-03-21 10 26110 0 2025-03-21 11 34415 3 2025-03-21 12 35940 3 2025-03-21 13 26516 1 2025-03-21 14 30908 3 2025-03-21 15 30916 4 2025-03-21 16 37333 1 2025-03-21 17 31202 3 2025-03-21 18 31201 1 2025-03-21 19 31767 1 2025-03-21 20 27635 1 2025-03-21 21 37939 1 2025-03-21 22 34318 3 2025-03-21 23 34372 1 2025-03-21 24 34634 4 2025-03-21 25 39385 2 2025-03-21 26 38318 3 2025-03-21 27 34336 0 2025-03-21 28 38340 3 2025-03-21 29 39309 3 2025-03-21 30 27200 0 2025-03-21 31 26960 4 2025-03-21 32 26670 3 2025-03-21 33 30920 0 2025-03-21 34 8224 0 2025-03-21 35 34224 3 2025-03-21 36 30948 1 2025-03-21 37 9723 3 2025-03-21 38 26750 1 2025-03-21 39 34630 0 2025-03-21 40 34409 1 2025-03-21 41 26330 1 2025-03-21 42 39307 3 2025-03-21 43 30949 3 2025-03-21 44 36957 0 2025-03-21 45 34610 1 2025-03-21 46 35941 0 2025-03-21 47 34310 0 2025-03-21 48 31768 1 2025-03-21 49 27550 1 2025-03-21 50 26935 1 2025-03-21 51 9729 1 2025-03-21 52 32902 3 2025-03-21 53 35953 3 2025-03-21 54 34407 0 2025-03-21 55 26700 1 2025-03-21 56 34206 1 2025-03-21 57 34309 0 2025-03-21 58 26537 1 2025-03-21 59 26620 1 2025-03-21 60 26790 1 2025-03-21 61 27990 1 2025-03-21 62 34327 3 2025-03-21 63 8385 3 2025-03-21 64 34223 3 2025-03-21 65 27300 3 2025-03-21 66 27720 1 2025-03-21 67 32801 4 2025-03-21 68 27400 0 2025-03-21 69 30932 0 2025-03-21 70 34426 1 2025-03-21 71 34319 3 2025-03-21 72 37938 1 2025-03-21 73 27450 1 2025-03-21 74 27610 2 2025-03-21 75 31840 3 2025-03-21 76 37969 0 2025-03-21 77 30927 4 2025-03-21 78 27165 1 2025-03-21 79 31003 0 2025-03-21 80 34296 1 2025-03-21 81 30930 3 2025-03-21 82 26109 1 2025-03-21 83 26630 1 2025-03-21 84 32905 1 2025-03-21 85 26450 1 2025-03-21 86 34325 4 2025-03-21 87 34654 1 2025-03-21 88 36244 1 2025-03-21 89 34242 1 2025-03-21 90 34011 1 2025-03-21 91 30572 0 2025-03-21 92 34411 1 2025-03-21 93 27640 1 2025-03-21 94 34320 0 2025-03-21 95 30913 0 2025-03-21 96 30951 3 2025-03-21 97 26910 3 2025-03-21 98 30917 0 2025-03-21 99 27630 1 2025-03-21 100 37941 1 2025-03-21 101 25121 1 2025-03-21 102 25016 1 2025-03-21 103 30954 3 2025-03-21 104 27150 0 2025-03-21 105 25325 1 2025-03-21 106 30923 3 2025-03-21 107 38329 3 2025-03-21 108 6678 1 2025-03-21 109 26950 1 2025-03-21 110 3132 3 2025-03-21 111 34385 3 2025-03-21 112 30933 1 2025-03-21 113 34003 0 2025-03-21 114 30902 0 2025-03-21 115 26870 1 2025-03-21 116 26925 3 2025-03-21 117 34392 0 2025-03-21 118 9721 1 2025-03-21 119 37940 3 2025-03-21 120 9391 1 2025-03-21 121 35952 1 2025-03-21 122 34627 1 2025-03-21 123 26536 1 2025-03-21 124 30950 3 2025-03-21 125 32819 0 2025-03-21 126 26895 1 2025-03-21 127 6676 0 2025-03-21 128 30997 0 2025-03-21 129 7150 4 2025-03-21 130 27100 0 2025-03-21 131 9726 3 2025-03-21 132 34410 1 2025-03-21 133 34605 1 2025-03-21 134 6742 1 2025-03-21 135 37358 3 2025-03-21 136 8298 0 2025-03-21 137 30953 3 2025-03-21 138 30995 3 2025-03-21 139 34321 3 2025-03-21 140 36243 1 2025-03-21 141 31408 1 2025-03-21 142 30201 3 2025-03-21 143 30936 0 2025-03-21 144 36930 3 2025-03-21 145 34001 0 2025-03-21 146 30996 3 2025-03-21 147 38333 3 2025-03-21 148 34329 1 2025-03-21 149 9728 1 2025-03-21 150 34222 4 2025-03-21 151 34632 4 2025-03-21 152 37381 3 2025-03-21 153 27625 1 2025-03-21 154 8540 3 2025-03-21 155 30909 3 2025-03-21 156 9724 1 2025-03-21 157 25067 1 2025-03-21 158 26985 3 2025-03-21 159 30555 0 2025-03-21 160 30901 1 2025-03-21 161 8967 3 2025-03-21 162 34335 3 2025-03-21 163 27500 1 2025-03-21 164 27930 1 2025-03-21 165 30928 0 2025-03-21 166 26760 1 2025-03-21 167 31608 4 2025-03-21

In [25]:
drop view final_tab;

Success: 16 rows affected

In [26]:
--Optional query to Join Original Clusters with Teradata
create volatile table final_tab as (
SELECT DISTINCT 
    hc.cluster_id AS Teradata_Cluster,
    original_cluster.cluster_id as Original_Cluster,
    hf.entity_id,      -- Include any other non-aggregated columns you want
    hf.entity_name,
    hf.count_admissions,
    hf.count_beds,
    hf.count_fte, 
    hf.sum_net_revenue, 
    hf.sum_opex,
    hf.score_case_mix
 FROM 
    v_features hf
JOIN 
    clusters hc ON hf.entity_id = hc.entity_id
LEFT JOIN original_cluster
    ON hf.entity_id = original_cluster.entity_id
)with data ON COMMIT PRESERVE ROWS;

Success: 0 rows affected

In [27]:
select * from final_tab

Teradata_Cluster Original_Cluster entity_id entity_name count_admissions count_beds count_fte sum_net_revenue sum_opex score_case_mix 1 3 2 3132 METHODIST HOSPITAL STONE OAK 26527 241 1100 363265494 262107239 1.61 2 2 1 27610 MISSION HOSPITAL 78804 821 4051 1368498651 987982056 1.88 3 4 1 26960 MEMORIAL HEALTH UMC SAVANNAH 53135 599 2655 769548229 607208617 1.88 4 1 0 6678 ST DAVID'S REHAB 0 5 0 2 26920 HCA HOUSTON NORTHWEST 24991 307 1368 423010643 332216940 1.56 6 3 2 8385 REGIONAL MEDICAL CENTER OF SAN JOSE 15363 192 1054 345167998 365409909 1.83 7 2 1 39385 METHODIST HOSPITAL 65761 827 4038 1371232053 964542580 1.84 8 4 1 31608 WESLEY MEDICAL CENTER 43153 506 2108 745045705 472300442 1.69 9 1 0 9729 RESEARCH PSYCHIATRIC CENTER 0 10 0 2 30913 HCA FLORIDA ORANGE PARK HOSPITAL 30231 348 1386 428004422 283528692 1.67 11 3 2 8540 RESTON HOSPITAL CENTER 23916 254 1079 427424085 296828574 1.69 12 4 1 32801 SUNRISE HOSPITAL AND MEDICAL CENTER 53776 777 3320 837504743 766218950 1.89 13 1 2 31408 TERRE HAUTE REGIONAL HOSPITAL 9785 208 473 113034255 112275744 1.56 14 0 2 30932 HCA FLORIDA LAWNWOOD HOSPITAL 34466 432 1637 531541178 357143367 1.72 15 3 2 8967 SOUTHERN HILLS HOSP & MED CTR-LAS VEGAS 30271 265 1091 348835620 259332579 1.56 16 4 1 34222 CENTENNIAL MEDICAL CENTER 46091 710 2855 1070659554 738754217 2.11 17 1 2 31767 GREENVIEW REGIONAL HOSPITAL 14437 118 438 143532991 109108076 1.54 18 0 2 30997 HCA FLORIDA OAK HILL HOSPITAL 32348 350 1275 370906368 246575422 1.52 19 3 2 9722 OVERLAND PARK REGIONAL MEDICAL CENTER 23839 271 1207 492537327 305214652 1.63 20 4 1 34325 MEDICAL CITY DALLAS 40471 803 2812 1087049498 685806996 1.63 21 1 2 34242 SOUTHERN HILLS MEDICAL CENTER 13424 115 506 150095045 110797255 1.70 22 0 2 34309 MEDICAL CITY ARLINGTON 29739 356 1373 471449023 292146310 1.58 23 3 2 9723 MENORAH MEDICAL CENTER 18173 166 821 293408942 221621917 1.71 24 4 1 34632 CJW MEDICAL CENTER 67338 763 2871 1080024474 701193821 1.68 25 1 4 26109 METHODIST TEXSAN HOSPITAL 10819 120 518 165664371 123692956 1.88 26 0 2 34392 METROPOLITAN METHODIST HOSPITAL 32942 317 1440 398222957 302810148 1.56 27 3 2 9726 CENTERPOINT MEDICAL CENTER 29090 285 1267 416841535 309036562 1.61 28 4 3 7150 RIVERSIDE COMMUNITY HOSPITAL 39112 513 2291 697036960 589547550 1.71 29 1 4 26700 CACHE VALLEY HOSPITAL 3815 28 127 35343014 31650025 1.04 30 0 3 6676 ST DAVID'S MEDICAL CENTER 32346 409 1780 621267278 417761678 1.73 31 3 2 26670 HCA FLORIDA CITRUS HOSPITAL 21267 204 784 248258233 176303370 1.74 32 4 3 30916 HCA FLORIDA NORTH FLORIDA HOSPITAL 52944 484 2039 657229814 432801468 1.59 33 1 5 6742 MEDICAL CITY LAS COLINAS 9753 72 388 108605216 80981148 1.41 34 0 3 8224 NORTH AUSTIN MEDICAL CENTER 41071 418 1689 539592175 373031490 1.36 35 3 2 26910 HCA HOUSTON TOMBALL 17901 241 887 228844142 192259193 1.44 36 4 3 30927 HCA FLORIDA OCALA HOSPITAL 59885 545 2220 782525111 513999395 1.76 37 1 5 9391 STONECREST MEDICAL CENTER 16664 115 561 197120950 122267790 1.40 38 0 3 8298 CORPUS CHRISTI MEDICAL CENTER 39350 421 1656 490845548 376009568 1.51 39 3 2 26985 HCA HOUSTON N CYPRESS 21429 163 1104 330903714 236889005 1.76 40 4 3 34634 HENRICO DOCTORS HOSPITAL 49427 503 1983 764179568 511933777 1.53 41 1 5 9721 LAFAYETTE REGIONAL HEALTH CENTER 4469 25 128 34683217 28837127 1.18 42 0 3 9727 RESEARCH MEDICAL CENTER 28051 442 1732 592154834 479487485 1.89 43 3 2 30201 ALASKA REGIONAL HOSPITAL 11554 179 867 365694336 232612364 1.96 44 1 5 9724 LEE'S SUMMIT MEDICAL CENTER 12824 85 484 168723329 120642981 1.59 45 0 3 26110 HCA FLORIDA MERCY HOSPITAL 29362 333 1324 380421391 287467643 1.43 46 3 2 30908 HCA FLORIDA WESTSIDE HOSPITAL 21373 246 988 325809380 223032684 1.84 47 1 5 9728 BELTON REGIONAL MEDICAL CENTER 11128 62 326 99673875 75927570 1.39 48 0 3 27100 SWEDISH MEDICAL CENTER 28983 349 1679 643709917 431666993 1.98 49 3 2 30909 HCA FLORIDA FORT WALTON-DESTIN HOSPITAL 28144 318 1129 352782387 246476358 1.63 50 1 5 25016 LONE PEAK HOSPITAL 12936 61 362 146960756 8928297

<p style = 'font-size:18px;font-family:Arial;color:#00233C'><b> 5. Model Accuracy Check</b></p>


<p style = 'font-size:16px;font-family:Arial;color:#00233C'><a href = 'https://docs.teradata.com/r/Enterprise_IntelliFlex_VMware/Database-Analytic-Functions/Model-Evaluation-Functions/TD_Silhouette'>TD_Silhouette</a> function is a method of interpretation and validation of consistency within clusters of data. The function determines how well the data is clustered among clusters.


<p style = 'font-size:16px;font-family:Arial;color:#00233C'>
Silhouette Score measures how well each point fits in its assigned cluster (range: -1 to 1).</p>
<p style = 'font-size:16px;font-family:Arial;color:#00233C'>
1 → Perfect clustering <br>
0 → Overlapping clusters <br>
-1 → Poor clustering (misclassified points) 
</p>

In [28]:

SELECT * FROM TD_Silhouette(
ON final_tab as InputTable
USING
 IdColumn('entity_id')
 ClusterIdColumn('Teradata_Cluster')
 TargetColumns('count_fte' , 'count_beds' ,'count_admissions' ,'score_case_mix','sum_net_revenue' ,'sum_opex')
 OutputType('CLUSTER_SCORES')
) as dt;

Teradata_Cluster silhouette_score 1 3 0.39522100358019274 2 2 0.9634131617466692 3 4 0.20367729590261954 4 1 0.6788511378554553 5 0 0.3334679240214497

<p style = 'font-size:16px;font-family:Arial;color:#00233C'>

Clusters 2 and 1 have very high silhouette scores (0.95), meaning they are well-separated and compact.<br>
Clusters 4 and 5 have relatively low scores (0.28 - 0.32), suggesting overlapping clusters. <br>
Overall average is moderate (0.4), indicating decent but not perfect clustering.

In [29]:
SELECT * FROM TD_Silhouette(
ON final_tab as InputTable
USING
 IdColumn('entity_id')
 ClusterIdColumn('Original_Cluster')
 TargetColumns('count_fte' , 'count_beds' ,'count_admissions' ,'score_case_mix','sum_net_revenue' ,'sum_opex')
 OutputType('CLUSTER_SCORES')
) as dt;

Original_Cluster silhouette_score 1 3 0.3157027259077065 2 5 0.20985513672869832 3 2 0.3055782736664925 4 4 -0.2567482861511197 5 1 0.28939898746592924

Cluster 4 has a negative silhouette score (0.256) → likely misclassified or overlapping significantly with other clusters. <br>
The highest score is  around 0.31, which is quite low overall, indicating poor cluster separation. <br>
Overall, the second model is weaker than the first one, as most clusters have lower scores compared to the first model.


Conclusions

1. The first K-Means model performs better overall (higher silhouette scores). <br>
2. Some clusters are well-formed (Cluster 1 and 2 with scores 0.95).<br>
3. However, Clusters 4 and 5 still need improvement (low scores).<br>
4. The second K-Means model has lower silhouette scores, with one cluster being misclassified (negative score).<br>
5. Indicates overlapping clusters and poor separation.<br>
6. Needs improvement in feature scaling, number of clusters (K), or different distance metrics.

Recommendation

1. Reevaluate the choice of K: Use the Elbow Method to determine the optimal number of clusters. <br>
2. Try different distance metrics: Instead of Euclidean distance, test cosine similarity or Manhattan distance.<br>
3. Feature Engineering: Remove redundant features or apply Principal Component Analysis (PCA) to improve separability.<br>
4. Check for outliers: Outliers can distort clustering results. Consider using DBSCAN if data has noise.